s3torchconnector/src/s3torchconnector/_s3dataset_common.py

```diff
@@ -41,22 +41,18 @@ def parse_s3_uri(uri: str) -> Tuple[str, str]:
 
 def get_objects_from_uris(
     object_uris: Union[str, Iterable[str]], client: S3Client
-) -> Iterable[S3Reader]:
+) -> Iterable[S3BucketKey]:
     if isinstance(object_uris, str):
         object_uris = [object_uris]
     # TODO: We should be consistent with URIs parsing. Revise if we want to do this upfront or lazily.
-    bucket_key_pairs = [parse_s3_uri(uri) for uri in object_uris]
+    bucket_key_pairs = [S3BucketKey(*parse_s3_uri(uri)) for uri in object_uris]
 
-    return bucket_key_pairs_to_objects(bucket_key_pairs, client)
+    return bucket_key_pairs
 
 
-def bucket_key_pairs_to_objects(
-    bucket_key_pairs: List[Tuple[str, str]], client: S3Client
-):
-    for bucket, key in bucket_key_pairs:
-        yield client.get_object(bucket, key)
-
-
-def list_objects_from_prefix(s3_uri: str, client: S3Client) -> S3BucketIterable:
+def list_objects_from_prefix(s3_uri: str, client: S3Client) -> Iterable[S3BucketKey]:
     bucket, prefix = parse_s3_uri(s3_uri)
-    return S3BucketIterable(client, bucket, prefix)
+    s3objects = list(S3BucketIterable(client, bucket, prefix))
+    bucket_key_pairs = [S3BucketKey(obj.bucket, obj.key) for obj in s3objects]
```

Comment:
We were deliberately returning an iterable here rather than a list. This uses less memory

Reply:
If we start returning a list here, would it eagerly fetch all the keys for IterableDataset? IterableDataset should fetch keys from S3 as we iterate over records. When we are returning an iterable thats the behaviour. If we convert it to list, we might change this behaviour.